frontend/catalyst/pennylane_extensions.py

```diff
@@ -167,13 +173,29 @@ def __call__(self, *args, **kwargs):
         )
 
         jaxpr = jax.make_jaxpr(self.fn)(*args)
-        if len(jaxpr.eqns) != 1:
-            raise TypeError("catalyst.grad can only be used on QNodes and other grad calls")
-        if jaxpr.eqns[0].primitive != jprim.func_p:
-            raise TypeError(
-                f"Attempting to differentiate something other than a function "
-                f"(got {jaxpr.eqns[0].primitive})"
-            )
+        assert len(jaxpr.eqns) == 1, "Expected jaxpr consisting of a single function call."
+        assert (
+            jaxpr.eqns[0].primitive == jprim.func_p
+        ), "Expected jaxpr consisting of a single function call."
+
+        if self.method != "fd":
+            qnode_jaxpr = jaxpr.eqns[0].params["call_jaxpr"]
+            return_ops = []
+            for res in qnode_jaxpr.outvars:
+                for eq in reversed(qnode_jaxpr.eqns):
```

Comment:
Why do we need this reversed?

Reply:
The original formulation is harder to read. My preference is just to remove the `reverse`. I'll revisit once you submit another patch.

Reply:
I added back the early exit 👍

Reply:
Cool! I didn't think too much about whether exit early was a possibility, but it makes sense.

Reply:
Looks like the early exit is what causes codecov to only award partial coverage on the for loop, annoying :/

https://app.codecov.io/gh/PennyLaneAI/catalyst/pull/75#3abdf3007845408f68a588566bfbf90d-R185

Reply:
Yeah, I was just looking into it. I guess the return is always the last statement? Is this an invariant in JAXPR?

Reply:
I don't think there is a return statement in JAXPR, only out variables which have been defined by one of the equations/statements. I'm pretty sure the test cases we have include return values that are not defined by the last equation though, so the `if` statement is fully covered on both branches.
It seems what is not fully covered on all branches is the loop statement 🤔

Reply:
I will add the exclusion comment on branch coverage for the for loop.